File: Java/044.java
import java.util.HashSet;
import java.util.Set;

public class Problem044 {

    // Function to check if a number is pentagonal
    public static boolean isPentagonal(long n) {
        // P_n = n(3n-1)/2
        // 2P_n = 3n^2 - n
        // 3n^2 - n - 2P_n = 0
        // Using quadratic formula: n = (1 +/- sqrt(1 - 4*3*(-2P_n))) / (2*3)
        // n = (1 +/- sqrt(1 + 24P_n)) / 6
        // For n to be an integer, (1 + 24P_n) must be a perfect square, and (1 + sqrt(1 + 24P_n)) must be divisible by 6.
        long discriminant = 1 + 24 * n;
        long sqrtDiscriminant = (long) Math.sqrt(discriminant);

        if (sqrtDiscriminant * sqrtDiscriminant != discriminant) {
            return false; // Not a perfect square
        }

        return (1 + sqrtDiscriminant) % 6 == 0;
    }

    public static void main(String[] args) {
        Set<Long> pentagonalNumbers = new HashSet<>();
        long minD = Long.MAX_VALUE;

        // Generate a sufficient number of pentagonal numbers
        // The problem implies that the numbers are not excessively large.
        // We can iterate and add pentagonal numbers to a set.
        // The upper limit for 'k' can be determined empirically or by analysis.
        // For this problem, a limit of around 3000-4000 for 'k' should be sufficient.
        for (int k = 1; k < 4000; k++) {
            pentagonalNumbers.add((long) k * (3 * k - 1) / 2);
        }

        for (long pj : pentagonalNumbers) {
            for (long pk : pentagonalNumbers) {
                if (pk <= pj) continue; // Ensure Pk > Pj

                long sum = pj + pk;
                long diff = pk - pj;

                if (pentagonalNumbers.contains(sum) && pentagonalNumbers.contains(diff)) {
                    if (diff < minD) {
                        minD = diff;
                    }
                }
            }
        }
        System.out.println(minD);
    }
}

File: Rust/044.rs
use std::collections::HashSet;

fn is_pentagonal(num: u32) -> bool {
    let test = 1 + 24 * num as u64;
    let sqrt_test = (test as f64).sqrt() as u64;
    if sqrt_test * sqrt_test != test {
        return false;
    }
    (1 + sqrt_test) % 6 == 0
}

fn main() {
    let mut pentagonal_numbers: Vec<u32> = Vec::new();
    let mut pentagonal_set: HashSet<u32> = HashSet::new();

    // Generate pentagonal numbers up to a certain limit
    // The problem implies that the numbers won't be excessively large.
    // A limit of 60,000 for n should be sufficient to find the answer.
    // P_60000 = 60000 * (3 * 60000 - 1) / 2 = 5,399,910,000
    // This is within u32 range.
    for n in 1..60000 {
        let p_n = n * (3 * n - 1) / 2;
        pentagonal_numbers.push(p_n);
        pentagonal_set.insert(p_n);
    }

    let mut min_diff = u32::MAX;

    // Iterate through pairs of pentagonal numbers
    for i in 0..pentagonal_numbers.len() {
        for j in (i + 1)..pentagonal_numbers.len() {
            let pj = pentagonal_numbers[i];
            let pk = pentagonal_numbers[j];

            let sum = pj + pk;
            let diff = pk - pj;

            if is_pentagonal(sum) && is_pentagonal(diff) {
                if diff < min_diff {
                    min_diff = diff;
                }
            }
        }
    }

    println!("{}", min_diff);
}
